src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/globalSelectionHeader.jsx

```diff
@@ -0,0 +1,336 @@
+      return true;
+    }
+
+    const nonDateKeys = ['projects', 'environments'];
+    // Update if selection values change
+    if (
+      !isEqual(
+        pick(this.props.selection, nonDateKeys),
+        pick(nextProps.selection, nonDateKeys)
+      ) ||
+      !isEqualWithDates(
+        pick(this.props.selection.datetime, DATE_TIME_KEYS),
+        pick(nextProps.selection.datetime, DATE_TIME_KEYS)
+      )
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.hasCustomRouting) {
+      return;
+    }
+
+    // If component has updated (e.g. due to re-render from a router action),
+    // update store values with values from router. Router should be source of truth
+    this.updateStoreIfChange(prevProps, this.props);
+  }
+
+  didQueryChange = (prevProps, nextProps) => {
+    const urlParamKeys = Object.values(URL_PARAM);
+    return !isEqual(
+      pick(prevProps.location.query, urlParamKeys),
+      pick(nextProps.location.query, urlParamKeys)
+    );
+  };
+
+  updateStoreIfChange = (prevProps, nextProps) => {
+    // Don't do anything if query parameters have not changed
+    //
+    // e.g. if selection store changed, don't trigger more actions
+    // to update global selection store (otherwise we'll get recursive updates)
+    if (!this.didQueryChange(prevProps, nextProps)) {
+      return;
+    }
+
+    const {
+      project,
+      environment,
+      period,
+      start,
+      end,
+      utc,
+    } = GlobalSelectionHeader.getStateFromRouter(nextProps);
+
+    if (start || end || period || utc) {
+      // Don't attempt to update date if all of these values are empty
+      updateDateTime({start, end, period, utc});
+    }
+
+    updateEnvironments(environment || []);
+    updateProjects(project || []);
+  };
+
+  // Returns `router` from props if `hasCustomRouting` property is false
+  getRouter = () => (!this.props.hasCustomRouting ? this.props.router : null);
+
+  handleChangeProjects = projects => {
+    this.setState({
+      project: projects,
+    });
+    callIfFunction(this.props.onChangeProjects, projects);
+  };
+
+  handleChangeEnvironments = environments => {
+    this.setState({
+      environment: environments,
+    });
+    callIfFunction(this.props.onChangeEnvironments, environments);
+  };
+
+  handleChangeTime = ({start, end, relative, utc}) => {
+    this.setState({start, end, period: relative, utc});
+    callIfFunction(this.props.onChangeTime, {start, end, relative, utc});
+  };
+
+  handleUpdateTime = () => {
+    let stateObjects = pick(this.state, DATE_TIME_KEYS);
+    if (!Object.values(stateObjects).some(i => i)) {
+      stateObjects = this.props.selection.datetime;
+    }
+    const {period, start, end, utc} = stateObjects;
+
+    const newValueObj = {
+      ...(defined(period) ? {period} : {start, end}),
+      utc,
+      zoom: null,
+    };
+
+    this.setState({
+      start: null,
```

Comment:
How about about keeping these grouped under `datetime`? (same format as global selection store)

Reply:
On second thoughts.... I guess this really doesn't really matter so feel free to ignore that comment